rclpy/src/rclpy/qos_events.cpp

```diff
@@ -0,0 +1,352 @@
+    rclpy_handle_get_pointer_from_capsule(pypublisher.ptr(), "rclpy_publisher_t"));
+  if (!wrapper) {
+    throw py::error_already_set();
+  }
+
+  rcl_publisher_event_type_t event_type =
+    rcl_publisher_event_type_from(pyevent_type);
+
+  unique_cstruct_ptr<rcl_event_t> event = create_zero_initialized_event();
+
+  rcl_ret_t ret = rcl_publisher_event_init(
+    event.get(), &(wrapper->publisher), event_type);
+  if (RCL_RET_OK != ret) {
+    if (RCL_RET_BAD_ALLOC == ret) {
+      rcl_reset_error();
+      throw std::bad_alloc();
+    }
+    if (RCL_RET_UNSUPPORTED == ret) {
+      throw UnsupportedEventTypeError("publisher event is unsupported");
+    }
+    throw RCLError("failed to create publisher event");
+  }
+
+  return event_wrap_in_capsule(std::move(event), pypublisher);
+}
+
+py::object
+subscription_event_create(py::object pyevent_type, py::capsule pysubscription)
+{
+  auto wrapper = static_cast<rclpy_subscription_t *>(
+    rclpy_handle_get_pointer_from_capsule(pysubscription.ptr(), "rclpy_subscription_t"));
+  if (!wrapper) {
+    throw py::error_already_set();
+  }
+
+  rcl_subscription_event_type_t event_type =
+    rcl_subscription_event_type_from(pyevent_type);
+
+  unique_cstruct_ptr<rcl_event_t> event = create_zero_initialized_event();
+
+  rcl_ret_t ret = rcl_subscription_event_init(
+    event.get(), &(wrapper->subscription), event_type);
+  if (RCL_RET_OK != ret) {
+    if (RCL_RET_BAD_ALLOC == ret) {
+      rcl_reset_error();
+      throw std::bad_alloc();
+    }
+    if (RCL_RET_UNSUPPORTED == ret) {
+      throw UnsupportedEventTypeError("subscription event is unsupported");
+    }
+    throw RCLError("failed to create subscription event");
+  }
+
+  return event_wrap_in_capsule(std::move(event), pysubscription);
+}
+
+}  // namespace
+
+py::object
+create_event(py::object pyevent_type, py::capsule pyparent)
+{
+  if (strcmp(pyparent.name(), "rclpy_subscription_t") == 0) {
+    return subscription_event_create(pyevent_type, pyparent);
+  }
+  if (strcmp(pyparent.name(), "rclpy_publisher_t") == 0) {
+    return publisher_event_create(pyevent_type, pyparent);
+  }
+  throw py::type_error("event parent is neither a publisher nor a subscription");
+}
+
+py::object
+take_event(py::capsule pyevent, py::capsule pyparent, py::object pyevent_type)
+{
+  auto event = static_cast<rcl_event_t *>(
+    rclpy_handle_get_pointer_from_capsule(pyevent.ptr(), "rcl_event_t"));
+  if (!event) {
+    throw py::error_already_set();
+  }
+
+  qos_event_data_filler_function * event_filler =
+    qos_event_data_filler_function_for(pyparent, pyevent_type);
+
+  qos_event_callback_data_t event_data;
+  rcl_ret_t ret = rcl_take_event(event, &event_data);
+  if (RCL_RET_OK != ret) {
+    if (RCL_RET_BAD_ALLOC == ret) {
+      rcl_reset_error();
+      throw std::bad_alloc();
+    }
+    if (RCL_RET_UNSUPPORTED == ret) {
+      rcl_reset_error();
+      throw std::runtime_error("take event is not implemented");
```

Comment:
I believe this isn't documented in `take_event`'s header.

Reply:
Indeed. This is nonsense though. There's no documentation of `rcl_take_event()` returning RCL_RET_UNSUPPORTED. There's no implementation that does that either. See 6ba21f2b61ad99b1909151869e8596e6cb8e1f7c for a fix.